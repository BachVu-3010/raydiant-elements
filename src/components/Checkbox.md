## Alignment
```jsx
initialState={a: true, b: false};
const linkToBool = require('./__demo__/linkToBool')(state, setState);
<div>
  <div>
    <span>Check the baseline alignment: </span>
    <Checkbox {...(linkToBool('a'))}>Checkbox</Checkbox>
  </div>
  <div style={{marginTop: '10px'}}>
    <div>Check the left alignment: </div>
    <Checkbox {...(linkToBool('b'))}>Checkbox</Checkbox>
  </div>
</div>
```
## In a row
```jsx
initialState={a: true, b: true, c: false, d: false};
const linkToBool = require('./__demo__/linkToBool')(state, setState);
<Row>
  <Checkbox {...(linkToBool('a'))}>I'm checked</Checkbox>
  <Checkbox {...(linkToBool('b'))}><em>I'm</em> checked too</Checkbox>
  <Checkbox {...(linkToBool('c'))}>I'm not checked</Checkbox>
  <Checkbox {...(linkToBool('d'))}><em>I'm</em> not checked either</Checkbox>
</Row>
```

## In a column
```jsx
initialState={a: true, b: true, c: false, d: false};
const linkToBool = require('./__demo__/linkToBool')(state, setState);
<Column>
  <Checkbox {...(linkToBool('a'))}>I'm checked</Checkbox>
  <Checkbox {...(linkToBool('b'))}><em>I'm</em> checked too</Checkbox>
  <Checkbox {...(linkToBool('c'))}>I'm not checked</Checkbox>
  <Checkbox {...(linkToBool('d'))}><em>I'm</em> not checked either</Checkbox>
</Column>
```

Disabled:
```jsx
<Checkbox checked disabled>Clicking on me won't do anything</Checkbox>
```

Narrow:
```jsx
initialState={a: true};
const linkToBool = require('./__demo__/linkToBool')(state, setState);
<div style={{maxWidth: '300px', border: '1px dashed #999'}}>
  <Checkbox {...linkToBool('a')}>Here's a fairly long label in a space that's not very wide. I sure hope this works out.</Checkbox>
</div>
```

Round:
```jsx
initialState = {a: true};
const linkToBool = require('./__demo__/linkToBool')(state, setState);
<Checkbox {...linkToBool('a')}/>
```