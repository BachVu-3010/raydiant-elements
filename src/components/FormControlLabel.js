// @flow
/* eslint-disable jsx-a11y/label-has-for */

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withStyles } from 'material-ui/styles';

export const styles = theme => ({
  root: {
    position: 'relative',
    display: 'inline-flex',
    alignItems: 'baseline',
    cursor: 'pointer',
    WebkitTapHighlightColor: theme.palette.common.transparent,
    ...theme.typography.body2,
  },
  disabled: {
    opacity: 0.5,
    cursor: 'default',
  },
  label: {
    marginLeft: theme.spacing.unit,
    userSelect: 'none',
  },
});

const propTypes = {
  className: PropTypes.string,
  control: PropTypes.element.isRequired,
  disabled: PropTypes.bool,
  inputRef: PropTypes.func,
  label: PropTypes.node,
};

const defaultProps = {
  className: '',
  disabled: false,
  inputRef: null,
  label: '',
  onChange: () => {},
};

const FormControlLabel = ({
  classes,
  className,
  control,
  disabled,
  inputRef,
  label,
}) => (
  <label
    className={classnames(
      classes.root,
      { [classes.disabled]: disabled },
      className,
    )}
  >
    <span>{React.cloneElement(control, { inputRef })}</span>
    <span className={classes.label}>{label}</span>
  </label>
);

FormControlLabel.propTypes = propTypes;
FormControlLabel.defaultProps = defaultProps;

export default withStyles(styles)(FormControlLabel);
