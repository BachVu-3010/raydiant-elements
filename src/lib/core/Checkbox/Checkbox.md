```js
initialState = {
  checked: false,
};

<Row inline>
  <Checkbox
    checked={state.checked}
    onChange={checked => setState({ checked })}
  />
  <Checkbox disabled checked={state.checked} />
  <Checkbox
    round
    checked={state.checked}
    onChange={checked => setState({ checked })}
  />
  <Checkbox round disabled checked={state.checked} />
</Row>;
```

### Labels

```js
initialState = {
  checked: false,
};

<Row inline>
  <Checkbox
    label="Label"
    checked={state.checked}
    onChange={checked => setState({ checked })}
  />
  <Checkbox label="Label" disabled checked={state.checked} />
  <Checkbox
    label="Label"
    round
    checked={state.checked}
    onChange={checked => setState({ checked })}
  />
  <Checkbox label="Label" round disabled checked={state.checked} />
</Row>;
```
