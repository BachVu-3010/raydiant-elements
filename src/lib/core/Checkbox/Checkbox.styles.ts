import { Theme } from '../../theme';
import { createStyles } from '../withStyles';

const styles = (theme: Theme) =>
  createStyles({
    root: {
      position: 'relative',
      display: 'inline-block',
      width: 20,
      height: 20,
    },
    checkbox: {
      // MUI Checkboxes have a width / height of 48px to compensate for the ripple effect
      // which adds undesirable padding to the element. As a workaround, we position it
      // absolutely inside a container element with the correct width / height.
      position: 'absolute',
      top: -14,
      left: -14,
      padding: 0,
      color: theme.checkbox.backgroundChecked,
      '&$checked': {
        color: theme.checkbox.backgroundChecked,
      },
    },
    container: {
      boxSizing: 'border-box',
      width: 20,
      height: 20,
      backgroundColor: theme.checkbox.background,
      border: `1px solid ${theme.checkbox.border}`,
      borderRadius: theme.shape.borderRadius,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    round: {
      borderRadius: '100%',
    },
    disabled: {
      background: 'transparent',
      borderColor: theme.checkbox.borderMuted,
    },
    containerChecked: {
      backgroundColor: theme.checkbox.backgroundChecked,
      borderColor: theme.checkbox.backgroundChecked,
    },
    icon: {
      height: '100%',
      width: '100%',
      color: theme.checkbox.foreground,
    },
  });

export default styles;
