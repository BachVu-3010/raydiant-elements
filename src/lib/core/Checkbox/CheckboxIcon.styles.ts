import { makeStyles, createStyles } from '../../styles';
import { Theme } from '../../theme';

export default makeStyles((theme: Theme) =>
  createStyles({
    indeterminate: {},

    icon: {
      color: theme.checkbox.foreground,

      '$round &': {
        color: '#fff',
      },

      '&$indeterminate': {
        color: theme.checkbox.background,
        width: '100%',
        height: '100%',
      },
    },
  }),
);
