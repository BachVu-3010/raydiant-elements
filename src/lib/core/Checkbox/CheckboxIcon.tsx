import * as React from 'react';
import cn from 'classnames';
import Check from '@material-ui/icons/Check';
import RoundIndeterminateCheck from '@material-ui/icons/IndeterminateCheckBoxRounded';
import IndeterminateCheck from '@material-ui/icons/IndeterminateCheckBoxSharp';
import useStyles from './CheckboxIcon.styles';

interface CheckIconProps {
  indeterminate?: boolean;
  round?: boolean;
}

const CheckIcon: React.FC<CheckIconProps> = ({
  indeterminate = false,
  round = false,
}) => {
  const classes = useStyles();
  const className = cn(classes.icon, {
    [classes.indeterminate]: indeterminate,
  });

  if (indeterminate) {
    if (round) {
      return (
        <RoundIndeterminateCheck className={className} fontSize="inherit" />
      );
    }
    return <IndeterminateCheck className={className} fontSize="inherit" />;
  }
  return <Check className={className} fontSize="inherit" />;
};

export default CheckIcon;
