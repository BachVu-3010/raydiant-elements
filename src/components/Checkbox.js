import React from 'react';
import PropTypes from 'prop-types';
import FormControlLabel from './FormControlLabel';
import CheckboxInput from './CheckboxInput';

const propTypes = {
  /** Child elements are used as the checkbox label. */
  children: PropTypes.node,
  /** Whether the checkbox is checked or not. */
  checked: PropTypes.bool,
  /** Whether the checkbox is disabled or not. */
  disabled: PropTypes.bool,
  /**
   * The ID of an element that labels this control.
   * Useful if this control is separated from its text.
  */
  labelledBy: PropTypes.string,
  /** The name of the checkbox */
  name: PropTypes.string,
  /** Called when the user clicks the control. */
  onChange: PropTypes.func,
  /** The value of the checkbox */
  value: PropTypes.string,
};
const defaultProps = {
  children: null,
  checked: false,
  disabled: false,
  labelledBy: null,
  name: null,
  onChange: () => {},
  value: '',
};

/**
 * Use for independent `true` or `false` values.
 *
 * Checkboxes without labels will be rendered as round, and are intended for selecting items
 * in a grid.
 */
const Checkbox = ({
  children,
  checked,
  disabled,
  labelledBy,
  name,
  onChange,
  value,
}) => (
  <FormControlLabel
    {...{ disabled }}
    control={
      <CheckboxInput
        {...{ checked, disabled, labelledBy, name, onChange, value }}
        variant={children || labelledBy ? null : 'round'}
      />
    }
    label={children}
  />
);

Checkbox.propTypes = propTypes;
Checkbox.defaultProps = defaultProps;

export default Checkbox;
