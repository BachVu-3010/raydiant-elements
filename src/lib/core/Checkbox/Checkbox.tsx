import MUIPaper from '@material-ui/core/Paper';
import cn from 'classnames';
import * as React from 'react';
import { testAttr } from '../../helpers';
import CheckboxIcon from './CheckboxIcon';
import useStyles from './Checkbox.styles';

interface CheckboxProps {
  checked: boolean;
  label?: React.ReactNode;
  indeterminate?: boolean;
  round?: boolean;
  disabled?: boolean;
  small?: boolean;
  onChange?: (checked: boolean) => void;
  onClick?: (e: React.MouseEvent<any>) => void;
  className?: string;
  testId?: string;
}

export const Checkbox: React.SFC<CheckboxProps> = ({
  className,
  checked,
  indeterminate= false,
  label,
  round,
  disabled,
  small,
  onClick,
  onChange,
  testId,
}) => {
  const classes = useStyles();

  const checkbox = (
    <MUIPaper
      className={cn(
        !label && className,
        classes.inputContainer,
        disabled && classes.disabled,
        round && classes.round,
        small && classes.small,
        indeterminate && classes.indeterminate,
      )}
      elevation={0}
    >
      <input
        type="checkbox"
        className={classes.input}
        checked={checked}
        disabled={disabled}
        onChange={e => {
          if (onChange) {
            onChange(e.target.checked);
          }
        }}
        onClick={!label ? onClick : undefined}
        {...testAttr(testId)}
      />
      <div className={cn(
        classes.checkbox,
        (checked || indeterminate) && classes.checked,
        indeterminate && classes.indeterminate,
      )}>
        {(checked || indeterminate) && <CheckboxIcon indeterminate={indeterminate} round={round} />}
      </div>
    </MUIPaper>
  );

  if (label) {
    return (
      <label
        className={cn(classes.labelContainer, disabled && classes.disabled, className)}
        onClick={onClick}
      >
        {checkbox}
        <div className={classes.label}>{label}</div>
      </label>
    );
  }

  return checkbox;
};

export default Checkbox;
