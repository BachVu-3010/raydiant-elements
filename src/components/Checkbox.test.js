import React from 'react';
import { shallow } from 'enzyme';
import Checkbox from './Checkbox';
import FormControlLabel from './FormControlLabel';
import toggleTest from '../test_helpers/toggle.test';

toggleTest(Checkbox);

test('Round', () => {
  const getVariant = c =>
    c.find(FormControlLabel).prop('control').props.variant;
  let component;
  // Checkbox without label -> round
  component = shallow(<Checkbox />);
  expect(getVariant(component)).toBe('round');
  // Checkbox with label -> default
  component = shallow(<Checkbox>Check it out</Checkbox>);
  expect(getVariant(component)).toBe(null);
  // Checkbox with labelledBy -> default
  component = shallow(<Checkbox labelledBy="foo" />);
  expect(getVariant(component)).toBe(null);
});
