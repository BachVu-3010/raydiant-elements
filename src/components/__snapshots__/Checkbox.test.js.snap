// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<label
  className="FormControlLabel-root-1"
>
  <span>
    <span
      className="MuiButtonBase-root-11 SwitchBase-checkbox-root-4 SwitchBase-checkbox-default-6 SwitchBase-checkbox-checked-7"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      role={undefined}
      tabIndex={null}
    >
      <input
        checked={true}
        className="SwitchBase-checkbox-input-5"
        disabled={false}
        name=""
        onChange={[Function]}
        type="checkbox"
        value=""
      />
      <span
        className="MuiTouchRipple-root-13"
      />
    </span>
  </span>
  <span
    className="FormControlLabel-label-3"
  >
    Yes or no
  </span>
</label>
`;
